CDAP-6154 Custom Namespace Mapping for Filesystem

rohitsinha54 opened the pull request:
Issue: https://issues.cask.co/browse/CDAP-6154
Build: http://builds.cask.co/browse/CDAP-DUT4347-1
ITN: http://builds.cask.co/browse/CDAP-ITM-5
## Summary
- Allow user to specify custom location to use for cdap namespace.
- Custom locations are expected to be absolute path. A check is performed to ensure this before creating namespace.
- If a custom location is provided we check for it to exists while creating namespace
- If a custom location was provided we don't delete it while deleting namespace.
- This works on standalone too. (Tested on unix system. It has not been tested on windows system)
- Currently there is no CachingNamespaceQueryAdmin for performance improvement. It will be added in a separate PR. 
- The NamespacedLocationFactory now has a NamespaceQueryAdmin which it uses to look up namespace meta to give the correct location which is namespaced.
- NamespacedLocationFactory now takes in RootLocationFactory rather than LocationFactory (whose basepath is cdap dir) since the custom mapping can be any absolute path and to work with these we need to work with root of the filesystem.
- Added NamespacedLocationFactoryTestClient which is same as earlier NamespacedLocationFactory this is helpful in unit tests as it does not perform a namespace meta lookup and hence the unit tests does not need to  create namespace to get namespaced locations. 
- Added a LocationUnitTestModule which gives the same Module as LocationRuntimeModule.getInMemory ones but binds the NamespacedLocationFactory to the above mentioned NamespacedLocationFactoryTestClient to be used in tests. 
- DatasetService, DatasetInstanceService and DatasetTypeHandler now uses NamespaceQueryAdmin rather than NamespaceStore which it was doing earlier to avoid circular dependency in Guice injection. 

TODO:
- [ ] Run integration tests on this branch
- [ ] Test more rigorously on a cluster for edge cases
- [ ] After Hbase Mapping is done fix the todo to clean the if statements.

anwar6953 commented:
Integration tests running against this branch:
http://builds.cask.co/browse/CDAP-ITM-5

rohitsinha54 commented:
@anwar6953 : Thank you.

anwar6953 commented on cdap-common/src/main/java/co/cask/cdap/common/guice/LocationUnitTestModule.java:
```diff
@@ -0,0 +1,46 @@
+ * {@link DefaultNamespacedLocationFactory} and hence in unit tests the namespace does not need to be created to get
+ * namespaces locations.
+ */
+public class LocationUnitTestModule {
```
Can this class be moved into src/test, instead of src/main ?

rohitsinha54 commented on cdap-common/src/main/java/co/cask/cdap/common/namespace/DefaultNamespacedLocationFactory.java:
```diff
@@ -47,7 +57,20 @@ public Location get(Id.Namespace namespaceId) throws IOException {
+    try {
+      hdfsDirectory = namespaceQueryAdmin.get(namespaceId).getConfig().getHdfsDirectory();
+    } catch (Exception e) {
+      Throwables.propagate(e);
```
@chtyim : Is this okay ? The namespaceQueryAdmin.get throws Exception from its existing signature but the underlying implementation can either throw IOException or NamespaceNotFoundException. To throw the exact exception from here we will need to change the signature there. But this also adds the overhead that everyone using namespacedLocationFactory.get have to handle NamespaceNotFoundException and its at a lot of places in unit tests. So I prefered to wrap it in runtime. Let me know.

anwar6953 commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java:
```diff
@@ -130,13 +130,17 @@ public synchronized void create(NamespaceMeta metadata) throws Exception {
 
+    // store the metadata first and then create namespaces in the storage handler.
+    // TODO (CDAP-6155): this will be switched back to original when we do hbase mapping where we will start passing the
+    // NamespaceMeta itself to the DatasetFramework. We should first create namespaces in underlying storage before
```
we don't need to pass the meta; only the config, right?

rohitsinha54 replied:
The config does not have the namespace if so I dont think its a good idea to pass  just the config. We use the id of namespace a lot of places here so we need that info.

rohitsinha54 commented:
ITN : http://builds.cask.co/browse/CDAP-ITM-6

bdmogal commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/StorageProviderNamespaceAdmin.java:
```diff
@@ -57,17 +59,31 @@ protected StorageProviderNamespaceAdmin(CConfiguration cConf, NamespacedLocation
+    // TODO (CDAP-6155): This is not clean. Once we do hbase mapping the DatasetFramework will pass down the
+    // NamespaceMeta to this create and we should look for the custom location in the config rather than
+    // checking like this.
+    if (namespaceHome.toString().endsWith(Joiner.on(File.separator).join(cConf.get(Constants.Namespace.NAMESPACES_DIR),
```
can't we instead pass the `NamespaceConfig` here somehow? String-matching doesn't seem fool-proof?

bdmogal replied:
oh nvm. just read the TODO above. I  know its temporary, but can you split this up? hard to read

rohitsinha54 replied:
Gokul has the PR open to pass the meta here. We are planning to merge  his PR before thing one so as soon as its merged I will change this to use the meta  than this string check.

bdmogal commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/StorageProviderNamespaceAdmin.java:
```diff
@@ -87,15 +103,25 @@ protected void create(Id.Namespace namespaceId) throws IOException, ExploreExcep
+    // TODO (CDAP-6155): This is not clean. Once we do hbase mapping the DatasetFramework will pass down the
+    // NamespaceMeta to this delete and we should look for the custom location in the config there rather than
+    // checking like this.
+    if (namespaceHome.toString().endsWith(Joiner.on(File.separator).join(cConf.get(Constants.Namespace.NAMESPACES_DIR),
```
reverse this condition and return, so you can reduce indentation

bdmogal replied:
hmmm but you have the explore step after this. ok. nvm again.

bdmogal commented on cdap-proto/src/main/java/co/cask/cdap/proto/NamespaceConfig.java:
```diff
@@ -28,14 +29,42 @@
+  @SerializedName("hive.database")
+  private final String hiveDatabase;
+
+  public NamespaceConfig(@Nullable String schedulerQueueName, @Nullable String hdfsDirectory, @Nullable String
```
nit: fix indentation by moving the entire hbaseNamespace declaration to the next line

bdmogal commented on cdap-proto/src/main/java/co/cask/cdap/proto/NamespaceConfig.java:
```diff
@@ -28,14 +29,42 @@
 
-  public NamespaceConfig(String schedulerQueueName) {
+  @SerializedName("hdfs.directory")
+  private final String hdfsDirectory;
```
can we just call this filesystemDirectory/fsDirectory, since we support other filesystems as well?

anwar6953 replied:
how about `rootDirectory`

rohitsinha54 replied:
rootDirectory it is !

bdmogal commented on cdap-proto/src/main/java/co/cask/cdap/proto/NamespaceConfig.java:
```diff
@@ -28,14 +29,42 @@
+  @SerializedName("hbase.namespace")
+  private final String hbaseNamespace;
+
+  @SerializedName("hive.database")
```
should we remove these for newer fields? @anwar6953

anwar6953 replied:
Yea - I guess it just depends what we want the user to see in the json that they pass in:
`hive.database`
or
`hiveDatabase`
I'm fine with either.

rohitsinha54 replied:
@bdmogal  @anwar6953  What needs to be done for this ? I like the serialized names more as user facing names.

rohitsinha54 commented:
ITN: http://builds.cask.co/browse/CDAP-ITM-8

gokulavasan commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/NamespaceHttpHandlerTest.java:
```diff
@@ -137,7 +138,10 @@ public void testCreateWithConfig() throws Exception {
     // prepare - create namespace with config in its properties
     String propertiesString = String.format("{\"%s\":\"%s\", \"%s\":\"%s\", \"%s\":%s}",
```
Is there a better way to generate this string? Escaping looks ugly. We might want to add more properties in the future.

chtyim replied:
Gson a map.

``` java
new Gson().toJson(ImmutableMap.of(
  ...
));
```

bdmogal replied:
this is not addressed?

rohitsinha54 commented:
ITN: http://builds.cask.co/browse/CDAP-ITM-10

rohitsinha54 commented on cdap-data-fabric/src/test/java/co/cask/cdap/data2/datafabric/dataset/service/StorageProviderNamespaceHandlerTest.java:
```diff
@@ -29,16 +34,24 @@
+    // Since the namespace directory already exist this creating the same should fail
+    Assert.assertEquals(500, createNamespace("myspace").getResponseCode());
+    // should be able to delete the namespace
     Assert.assertEquals(200, deleteNamespace("myspace").getResponseCode());
```
Just realized I can add more unit test here specifically which test custom namespace mapping dir does not get create/deleted. Will do so.

anwar6953 commented on cdap-data-fabric/src/test/java/co/cask/cdap/data2/datafabric/dataset/service/StorageProviderNamespaceHandlerTest.java:
```diff
@@ -29,16 +34,24 @@
  */
 public class StorageProviderNamespaceHandlerTest extends DatasetServiceTestBase {
 
+  private static final Gson GSON = new GsonBuilder().create();
```
simply can do new Gson()

anwar6953 commented on cdap-proto/src/main/java/co/cask/cdap/proto/NamespaceConfig.java:
```diff
@@ -46,6 +46,10 @@ public NamespaceConfig(@Nullable String schedulerQueueName, @Nullable String roo
   }
 
+  public NamespaceConfig() {
+    this(null, null, null, null);
```
does the scheduler queue name support a value of null? (instead of empty string)?

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java:
```diff
@@ -305,7 +306,14 @@ private BodyConsumer deployApplication(final HttpResponder responder,
+    } catch (NamespaceNotFoundException e) {
+      responder.sendString(HttpResponseStatus.NOT_FOUND, e.getMessage());
+    } catch (UnauthenticatedException e) {
+      responder.sendString(HttpResponseStatus.FORBIDDEN, e.getMessage());
```
Why need to catch explicitly? Isn't that we have an exception handler that will inspect the exception and responds appropriately?

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java:
```diff
@@ -113,8 +114,25 @@ public void create(HttpRequest request, HttpResponder responder, @PathParam("nam
       NamespaceConfig config = metadata.getConfig();
-      if (config != null && !Strings.isNullOrEmpty(config.getSchedulerQueueName())) {
-        builder.setSchedulerQueueName(config.getSchedulerQueueName());
+      if (config != null) {
```
Why this logic is in the handler? Should the namespace admin take a `NamespaceConfig` and handle it appropriately? E.g. in future, when you add validations of the provided HBase/Hive namespace, it shouldn't be done in here.

rohitsinha54 replied:
The  logic of handling the config is in Admin. Here we are just creating the NamespaceMeta from what was  provided. Since the Handler takes the  id of namespace and metadata as body and recreates the meta with the  namespace id coming from the path param we are just doing that here. Does it looks wrong ?

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java:
```diff
@@ -625,8 +627,13 @@ public ArtifactDetail write(final Id.Artifact artifactId,
+      fileDirectory = namespacedLocationFactory.get(artifactId.getNamespace(),
+                                                    ARTIFACTS_PATH).append(artifactId.getName());
+    } catch (NamespaceNotFoundException | UnauthenticatedException e) {
+      throw Throwables.propagate(e);
```
The only caller to this method is from `ArtifactRepository` and in there the method actually has `throws Exception`. I think it's better to have this method just `thorws Exception` as well instead of wrapping it.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/namespace/DefaultNamespacedLocationFactory.java:
```diff
@@ -31,23 +34,42 @@
+  public Location get(Id.Namespace namespaceId, @Nullable String subPath) throws IOException,
+    NamespaceNotFoundException, UnauthenticatedException {
+    String hdfsDirectory = null;
+    hdfsDirectory = namespaceQueryAdmin.get(namespaceId).getConfig().getRootDirectory();
```
Merge line 63 and 64.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceQueryAdmin.java:
```diff
@@ -56,7 +58,8 @@ public DefaultNamespaceQueryAdmin(NamespaceStore nsStore) {
   @Override
-  public NamespaceMeta get(Id.Namespace namespaceId) throws Exception {
+  public NamespaceMeta get(Id.Namespace namespaceId) throws NamespaceNotFoundException, IOException,
+    UnauthenticatedException {
```
When will it be thrown? I don't see the operation can result in this. Also in this PR I see a lot of signatures has this exception added, why? When will it be thrown? Many of those calls are intra-CDAP service calls, which shouldn't have this exception, isn't it?

anwar6953 replied:
We should update the code to have a handler in app fabric (or some system container) that specifically serves this GET request.

rohitsinha54 replied:
@anwar6953 : That change is needed for impersonation right ? Lets do it in an independent PR.

anwar6953 commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java:
```diff
@@ -113,8 +114,25 @@ public void create(HttpRequest request, HttpResponder responder, @PathParam("nam
+          }
+          builder.setRootDirectory(config.getRootDirectory());
+        }
+        if (!Strings.isNullOrEmpty(config.getHbaseNamespace())) {
```
need to apply a similar change in DefaultNamespaceAdmin#updateProperties.

anwar6953 replied:
unless you don't support updating these properties, then you should have code there warning the user, if they try.

Commit:
CDAP-6154 Custom Namespace Mapping for Filesystem

Fix unit test

Add a new unit test for custom location

store namespacemeta first before creating underlying namespaces

Use NamesapceAdmin rather than NamespaceStore in StreamFileJanitor

Fix Explore Unit tests

Use RootLocationFactory only for custom mapping

fix InMemoryExploreServiceTest

Move the config checks in Admin

Do not allow root directory to be updated

Fix app-fabric-tests

fix cdap-master tests

Fix  comments and unit tests

Don't lookup namespace meta for system ns

Address comments

Address comments and fix integration tests

Addressed comments

Propagate NamespaceNotFoundException without wrapping into IOException

rohitsinha54 commented:
@bdmogal : I looked up from LocalMRJobInfoFetcherTest and it doesn't do that. Aren't all these gone once the unit test finishes ?

bdmogal commented:
yes, but tests should always clean up. If LocalMRJobInfoFetcherTest is not cleaning up, we should fix it.

bdmogal commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/namespace/StorageProviderNamespaceAdminTest.java:
```diff
@@ -76,5 +84,31 @@ public void test() throws Exception {
+    try {
+      storageProviderNamespaceAdmin.create(myspaceMeta);
+      Assert.fail("Expected exception to be thrown while creating namespace earlier");
+    } catch (Exception e) {
```
just expect `IOException` instead of using `instanceof`?

bdmogal commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/namespace/StorageProviderNamespaceAdminTest.java:
```diff
@@ -76,5 +84,31 @@ public void test() throws Exception {
     Assert.assertFalse(namespaceLocation.exists());
+
+    // test for custom location mapped namespace
+    myspace = new NamespaceId("custom");
```
Let's move this to its own test. Seems like separate functionality, and the cost of setup for these tests is not much, since we don't even deploy an application here.

bdmogal commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/namespace/StorageProviderNamespaceAdminTest.java:
```diff
@@ -76,5 +84,31 @@ public void test() throws Exception {
+    // now create the home directory too
+    // create the custom location
+    Location customLocation = locationFactory.create("custom");
+    customLocation.mkdirs();
```
let's move the creation after the `storageProviderNamespaceAdmin.create(myspaceMeta)` statement. That way, after creating the namespace via SPNA, you can assert that the `Location` does not exist. Then create it manually, followed by delete namespace.

bdmogal commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/namespace/StorageProviderNamespaceAdminTest.java:
```diff
@@ -76,5 +84,31 @@ public void test() throws Exception {
+    storageProviderNamespaceAdmin.create(myspaceMeta);
+    storageProviderNamespaceAdmin.delete(myspace);
+    // namespace location should still exists
+    Assert.assertTrue(customLocation.exists());
```
nit: should delete `customLocation` after this, just for completeness

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java:
```diff
@@ -134,13 +134,18 @@ public synchronized void create(NamespaceMeta metadata) throws Exception {
     }
 
+    // store the meta first in the namespace store because namespacedlocationfactory need to look up location
+    // mapping from namespace config
```
Why? I don't think this ordering is right.

anwar6953 replied:
I think its because of how the NamespacedLocationFactory is designed/implemented.
It will query the namespace store, to lookup the mapping (to know whether to have custom mapping or not).

anwar6953 replied:
However, we can avoid this lookup to the store, simply because the handler (and the namespace admin doing the creating) already knows whether there's a custom mapping or not.
So we don't need to make this ordering change here, right?

rohitsinha54 replied:
NamespaceAdmin knows if there is ordering or not if the NamespacedLocationFactory does not determine this on its own then this information need to be passed down to it. The current get of NamespacedLocationFactory does not have a parameter to say this if we change it  that means all the classes calling it will need to do the lookup on their own and pass down whether there is custom mapping or not. This seems weird since this should be encapsulated inside the NamespacedLocationFactory. Unless we add another method like create in namespaced location factory which is used only by the NamespaceAdmin while creating the namespaceadmin which does not do lookup on its own but depend on the called to tell if there is mapping or not. Should we do that ?

anwar6953 replied:
Can't we simply avoid calling the NamespacedLocationFactory, if there's no custom mapping?

rohitsinha54 replied:
Looked into it carefully. We can avoid calling NamespacedLocationFactory if there is custom mapping but when there is no custom mapping we call it with namespace id and NamespacedLocationFactory.get looks up mapping on its own to see what location to give.

bdmogal replied:
@rohitsinha54 when there is no custom mapping why does NamespacedLocationFactory need to lookup the mapping to find the location?

rohitsinha54 replied:
To know that there is no  custom mapping. It need to figure it out somehow unless we pass down this information to it from NamespaceAdmin through parameter or  something. That will require adding another get() and take like a boolean which says customMapping or not. But doing this does not look clean to me.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/AbstractStorageProviderNamespaceAdmin.java:
```diff
@@ -88,15 +115,26 @@ public void create(NamespaceMeta namespaceMeta) throws IOException, ExploreExcep
+    try {
+      if (Strings.isNullOrEmpty(namespaceQueryAdmin.get(namespaceId.toId()).getConfig().getRootDirectory())) {
+        if (namespaceHome.exists()) {
+          if (!namespaceHome.delete(true)) {
```
combine the two `if` into one.

rohitsinha54 commented:
New build: http://builds.cask.co/browse/CDAP-DUT4347-23

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/AbstractStorageProviderNamespaceAdmin.java:
```diff
@@ -57,18 +66,36 @@
+      // since this is a custom location we expect it to exists
+      if (!customLocation.exists()) {
+        throw new IOException(String.format("The provided home directory '%s' for namespace '%s' does not exists. " +
+                                              "Please create it on filesystem with sufficient privileges for the user" +
```
good to capture the user in this message? I know its not part of your PR though, so let's add a TODO?

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/AbstractStorageProviderNamespaceAdmin.java:
```diff
@@ -57,18 +66,36 @@
+      }
+      // since this is a custom location we expect it to exists
+      if (!customLocation.exists()) {
+        throw new IOException(String.format("The provided home directory '%s' for namespace '%s' does not exists. " +
```
also, nit: if you move the `String.format` or even just the message to the next line, this may become much easier to read.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/AbstractStorageProviderNamespaceAdmin.java:
```diff
@@ -88,15 +115,19 @@ public void create(NamespaceMeta namespaceMeta) throws IOException, ExploreExcep
-        throw new IOException(String.format("Error while deleting home directory '%s' for namespace '%s'",
-                                            namespaceHome, namespaceId.getNamespace()));
+    try {
+      if (Strings.isNullOrEmpty(namespaceQueryAdmin.get(namespaceId.toId()).getConfig().getRootDirectory())) {
```
can you break this up? hard to read.

bdmogal replied:
It could also be useful to move this to its own `private boolean isCustomRootDirectory`, since its used twice

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/AbstractStorageProviderNamespaceAdmin.java:
```diff
@@ -88,15 +115,19 @@ public void create(NamespaceMeta namespaceMeta) throws IOException, ExploreExcep
+                                              namespaceHome, namespaceId));
+        }
+      } else {
+        LOG.debug("Custom location mapping %s was found while deleting namespace %s. Skipping to delete the location.",
```
"Skipping location delete".

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DistributedStorageProviderNamespaceAdmin.java:
```diff
@@ -41,10 +42,11 @@
-                                           NamespacedLocationFactory namespacedLocationFactory,
-                                           ExploreFacade exploreFacade, HBaseTableUtil tableUtil) {
-    super(cConf, namespacedLocationFactory, exploreFacade);
+  public DistributedStorageProviderNamespaceAdmin(CConfiguration cConf,
```
why public?

bdmogal replied:
doesn't seem like its used anywhere. please revert this

rohitsinha54 replied:
hmm weird. Looks like intellij did it.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/LocalStorageProviderNamespaceAdmin.java:
```diff
@@ -27,8 +28,8 @@
-  LocalStorageProviderNamespaceAdmin(CConfiguration cConf, NamespacedLocationFactory namespacedLocationFactory,
-                                     ExploreFacade exploreFacade) {
-    super(cConf, namespacedLocationFactory, exploreFacade);
+  public LocalStorageProviderNamespaceAdmin(CConfiguration cConf, NamespacedLocationFactory namespacedLocationFactory,
```
ditto

rohitsinha54 replied:
Surely itellij did  it

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/AbstractStorageProviderNamespaceAdmin.java:
```diff
@@ -39,12 +43,17 @@
-  AbstractStorageProviderNamespaceAdmin(CConfiguration cConf, NamespacedLocationFactory namespacedLocationFactory,
-                                        ExploreFacade exploreFacade) {
+
+  protected AbstractStorageProviderNamespaceAdmin(CConfiguration cConf,
```
this is also a fairly internal class. why protected.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java:
```diff
@@ -269,7 +274,10 @@ public synchronized void updateProperties(Id.Namespace namespaceId, NamespaceMet
     }
-
+    if (config != null && !Strings.isNullOrEmpty(config.getRootDirectory())) {
+      throw new IllegalArgumentException(String.format("%s is immutable and cannot be updated.",
```
let's not use immutable, since that is a very software engineering term. Just "cannot be updated" is fine. Also, instead of `IllegalArgumentException` here, you should throw `BadRequestException` Are there tests for this?

rohitsinha54 replied:
Yes sir, there is a test for it.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java:
```diff
@@ -616,8 +616,9 @@ public ArtifactDetail write(final Id.Artifact artifactId,
-    Location fileDirectory =
-      namespacedLocationFactory.get(artifactId.getNamespace(), ARTIFACTS_PATH).append(artifactId.getName());
+    Location fileDirectory = null;
+      fileDirectory = namespacedLocationFactory.get(artifactId.getNamespace(),
```
why? unintended change?

rohitsinha54 replied:
intellij

bdmogal commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdminTest.java:
```diff
@@ -107,6 +109,22 @@ public void testNamespaces() throws Exception {
+
+    // check that root directory for a namespace cannot be updated
+    // create the custom directory
+    File customspaceDir = tmpFolder.newFolder("customspace");
```
it seems like you could split this into a separate test?

bdmogal commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/namespace/StorageProviderNamespaceAdminTest.java:
```diff
@@ -45,24 +53,78 @@
   }
+
+  @Test
+  public void testNSWithCustomLocation() throws ExploreException, SQLException, IOException {
```
nit: if the above test just throws `Exception`, maybe clean it up here too?

bdmogal commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/namespace/StorageProviderNamespaceAdminTest.java:
```diff
@@ -45,24 +53,78 @@
+    try {
+      storageProviderNamespaceAdmin.create(customSpaceMeta);
+      Assert.fail("Expected exception to be thrown while creating namespace with custom location since the custom " +
+                    "location does not exists at this point.");
```
nit: does not exist

bdmogal commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/namespace/StorageProviderNamespaceAdminTest.java:
```diff
@@ -45,24 +53,78 @@
+    storageProviderNamespaceAdmin.delete(customSpace);
+    // custom namespace location should still exists
+    Assert.assertTrue(custom.exists());
+    custom.delete();
```
should delete stuff from `nsStore` as well, so there are no side-effects?

bdmogal commented on cdap-common/src/main/java/co/cask/cdap/common/namespace/DefaultNamespacedLocationFactory.java:
```diff
@@ -47,13 +59,46 @@ public Location get(Id.Namespace namespaceId) throws IOException {
-    Location namespaceLocation = locationFactory.create(namespaceDir).append(namespaceId.getId());
+    String rootDirectory;
+    Location namespaceLocation;
+    if (Id.Namespace.DEFAULT == namespaceId || Id.Namespace.SYSTEM == namespaceId || Id.Namespace.CDAP == namespaceId) {
```
ummm... object comparison using `==`? Better use `equals`

rohitsinha54 replied:
issssh! My bad!

bdmogal commented on cdap-common/src/main/java/co/cask/cdap/common/namespace/DefaultNamespacedLocationFactory.java:
```diff
@@ -47,13 +59,46 @@ public Location get(Id.Namespace namespaceId) throws IOException {
+   *
+   * @param namespaceId the simple cdap namespace
+   * @return location of the namespace
+   * @throws IOException
```
remove

bdmogal commented on cdap-common/src/main/java/co/cask/cdap/common/namespace/NamespaceQueryAdmin.java:
```diff
@@ -50,3 +50,4 @@
    */
   boolean exists(Id.Namespace namespaceId) throws Exception;
 }
+
```
remove

bdmogal commented on cdap-common/src/main/java/co/cask/cdap/common/namespace/NamespacedLocationFactoryTestClient.java:
```diff
@@ -0,0 +1,64 @@
+  private final String namespaceDir;
+
+  @Inject
+  public NamespacedLocationFactoryTestClient(CConfiguration cConf, LocationFactory locationFactory) {
```
needs to be public?

rohitsinha54 replied:
ya in some test I am creating it to bind to specifically this instance

bdmogal commented on cdap-common/src/main/java/co/cask/cdap/common/namespace/NamespacedLocationFactoryTestClient.java:
```diff
@@ -0,0 +1,64 @@
+ * perform lookup for {@link NamespaceMeta} like {@link DefaultNamespacedLocationFactory} does and hence allow unit
+ * tests to use it without creating namespace meta for the namespace.
+ */
+public class NamespacedLocationFactoryTestClient implements NamespacedLocationFactory {
```
move this to `src/test` if it is only used in tests

bdmogal commented on cdap-data-fabric-tests/src/test/java/co/cask/cdap/data/stream/StreamFileTestUtils.java:
```diff
@@ -36,7 +36,7 @@ private StreamFileTestUtils() {
   public static Location getStreamBaseLocation(NamespacedLocationFactory namespacedLocationFactory,
                                                Id.Stream streamId) throws IOException {
-    return namespacedLocationFactory.get(streamId.getNamespace()).append(streamId.getId());
+      return namespacedLocationFactory.get(streamId.getNamespace()).append(streamId.getId());
```
revert change

bdmogal commented on cdap-data-fabric/src/test/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFrameworkTest.java:
```diff
@@ -155,8 +154,13 @@ public void onChange(ServiceDiscovered serviceDiscovered) {
     startLatch.await(5, TimeUnit.SECONDS);
 
+    // the framework.delete looks up namespace config through namespaceadmin add the meta there too.
+    NamespaceMeta systemMeta = new NamespaceMeta.Builder().setName(Id.Namespace.SYSTEM).build();
```
can you create a method called `createNamespace`, to do whatever it takes to simulate a namespace, and call it for two namespaces?

bdmogal replied:
also shouldn't location factory stuff happen automatically since you're using namespaceadmin?

rohitsinha54 replied:
the namespaceadmin here is just in-mem one so we need to create it explicitly. binding the default one needs a lot of dependencies which we dont have here.

Create the namespace create and delete methods.

bdmogal commented on cdap-data-fabric/src/test/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFrameworkTest.java:
```diff
@@ -188,9 +192,13 @@ public void testSystemNamespace() throws DatasetManagementException {
+  public void after() throws Exception {
+    // since we stored namespace meta through admin so that framework.delete can lookup namespaceconfig clean the
+    // meta from there too
     namespacedLocationFactory.get(NAMESPACE_ID).delete(true);
```
if you're using NamespaceAdmin, shouldn't this happen automatically?

bdmogal commented on cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java:
```diff
@@ -1294,11 +1294,11 @@ private String getHiveDatabase(@Nullable String namespace) {
    * @throws ExploreException
    */
-  protected Map<String, String> startSession() throws IOException, ExploreException {
+  protected Map<String, String> startSession() throws Exception {
```
why this change?

rohitsinha54 replied:
changed this because i made schedulerQueueResolver.getQueue to throw exception

anwar6953 replied:
schedulerQueueResolver.getQueue shouldn't throw Exception. It should only need to throw IOException.

bdmogal commented on cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java:
```diff
@@ -1294,11 +1294,11 @@ private String getHiveDatabase(@Nullable String namespace) {
   }
 
-  protected Map<String, String> startSession(Id.Namespace namespace) throws IOException, ExploreException {
+  protected Map<String, String> startSession(Id.Namespace namespace) throws Exception {
```
ditto

rohitsinha54 replied:
changed this because I  made schedulerQueueResolver.getQueue to throw exception

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetServiceManager.java:
```diff
@@ -53,7 +54,7 @@
  * Note: This should not be used outside upgrade tool.
  */
-public class DatasetServiceManager extends AbstractIdleService {
+public class DatasetServiceManager  extends AbstractIdleService {
```
revert change

bdmogal commented on cdap-proto/src/main/java/co/cask/cdap/proto/NamespaceConfig.java:
```diff
@@ -19,23 +19,58 @@
+  public static final String HIVE_DATABASE = "hive.database";
+
+  @SerializedName(SCHEDULER_QUEUE_NAME)
   private final String schedulerQueueName;
```
let's make this non-nullable again so it is backward-compatible?

rohitsinha54 replied:
Just to clarify, I have reverted the changes to this. This comment still shows up because this line doesn't need to change for this.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/AbstractStorageProviderNamespaceAdmin.java:
```diff
@@ -57,18 +66,36 @@
-                                            namespaceHome, namespaceMeta.getName()));
+    Location namespaceHome;
+    String rootDirectory = namespaceMeta.getConfig().getRootDirectory();
+    if (Strings.isNullOrEmpty(rootDirectory)) {
```
see https://github.com/caskdata/cdap/pull/6068/files#r71074638. This class is soon going to be unmanageable. We should move the FS stuff to its own private create/delete methods at the very least.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/AbstractStorageProviderNamespaceAdmin.java:
```diff
@@ -67,8 +67,24 @@ protected AbstractStorageProviderNamespaceAdmin(CConfiguration cConf,
     Location namespaceHome;
-    String rootDirectory = namespaceMeta.getConfig().getRootDirectory();
-    if (Strings.isNullOrEmpty(rootDirectory)) {
+    if (hasCustomLocation(namespaceMeta)) {
```
@rohitsinha54 lets move the filesystem stuff to its own create/delete methods. this code is going to get unmanageable soon, after HBase and Hive stuff is added

bdmogal commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdminTest.java:
```diff
@@ -122,7 +129,7 @@ public void testNamespaces() throws Exception {
                                       new NamespaceMeta.Builder(nsMeta).setRootDirectory("/newloc").build());
       Assert.fail();
-    } catch (IllegalArgumentException e) {
+    } catch (BadRequestException e) {
```
nit: expect this exception instead in the `@Test` annotation

bdmogal commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/NamespaceHttpHandlerTest.java:
```diff
@@ -138,10 +139,17 @@ public void testCreateWithConfig() throws Exception {
+
+    String propertiesString = GSON.toJson(ImmutableMap.of(NAME_FIELD, NAME, DESCRIPTION_FIELD, DESCRIPTION,
+                                                          CONFIG_FIELD,
+                                                          ImmutableMap.of(
```
nit: can you break this to make it more readable? Perhaps make the config a separate variable?

bdmogal commented:
almost there. 4 more comments.

anwar6953 commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java:
```diff
@@ -269,7 +275,9 @@ public synchronized void updateProperties(Id.Namespace namespaceId, NamespaceMet
       builder.setSchedulerQueueName(config.getSchedulerQueueName());
     }
-
+    if (config != null && !Strings.isNullOrEmpty(config.getRootDirectory())) {
```
We should also throw an exception if someone tries to remove the field.
For instance, if the value is empty string or null in the request, but currently is some custom mapping.
Right?

anwar6953 replied:
Instead of checking if its non-null and non-empty, I think we should check inequality with the existing value.

rohitsinha54 replied:
Good idea!

anwar6953 commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdminTest.java:
```diff
@@ -107,6 +109,22 @@ public void testNamespaces() throws Exception {
+
+    // Updating the root directory for a namespace should fail
+    try {
+      namespaceAdmin.updateProperties(nsMeta.getNamespaceId().toId(),
```
also have a test case for when user tries to remove the custom mapping.

anwar6953 commented:
I think cdap will fail to startup because explore twill runnable doesn't have bindings for NamespaceQueryHandler.
If it does indeed fail, you can just add it like so:
https://github.com/caskdata/cdap/blob/bb98e8f6c1a7af4a84f5d7203039ba25574a5028/cdap-master/src/main/java/co/cask/cdap/data/runtime/main/ExploreServiceTwillRunnable.java#L99

rohitsinha54 commented:
@anwar6953 : CDAP Starts up it has been tested numerous times and there is binding for NamespaceQueryAdmin for ExploreServiceTwillRunnable. Its coming from NamespaceClientRuntimeModule.

rohitsinha54 commented:
Build: http://builds.cask.co/browse/CDAP-DUT4347-27
ITN: http://builds.cask.co/browse/CDAP-ITM-15

russorat commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/AbstractStorageProviderNamespaceAdmin.java:
```diff
@@ -86,21 +84,70 @@ public void create(NamespaceMeta namespaceMeta) throws IOException, ExploreExcep
-      LOG.warn(String.format("Home directory '%s' for namespace '%s' does not exist.",
-                             namespaceHome, namespaceId));
+    } catch (Exception e) {
+      throw new IOException(String.format("Error while deleting home directory %s for namespace %s ", namespaceHome,
```
sometimes you surround the %s with single quotes, sometimes you don't. is that on purpose?

rohitsinha54 replied:
I generally don't do it some people do it.

bdmogal commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdminTest.java:
```diff
@@ -125,13 +125,11 @@ public void testConfigUpdateFailures() throws Exception {
+                                    new NamespaceMeta.Builder(nsMeta).setRootDirectory("/newloc").build());
+
+    // removing the root directory mapping for a namespace should fail
+    namespaceAdmin.updateProperties(nsMeta.getNamespaceId().toId(), new NamespaceMeta.Builder(nsMeta).build());
```
this won't work, since the test will exit at line 128. If you want to test multiple such scenarios, then you should `try...catch` on line 128.

bdmogal commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/NamespaceHttpHandlerTest.java:
```diff
@@ -141,14 +141,13 @@ public void testCreateWithConfig() throws Exception {
     // prepare - create namespace with config in its properties
 
 
+    ImmutableMap<String, String> namespaceConfigString = ImmutableMap.of(NamespaceConfig.SCHEDULER_QUEUE_NAME,
```
just Map

anwar6953 commented:
I see. I didn't notice you added the binding to `ExploreServiceTwillRunnable` already.

anwar6953 commented:
last commit looks fine to me.

bdmogal commented:
Restarted Build: http://builds.cask.co/browse/CDAP-DUT4347-28

bdmogal commented:
Integration tests have known failures. As long as we have a green unit test build, this PR is good to go.

Commit:
Check existense of location from locationfactory

rohitsinha54 commented:
Tested on secure cluster with FileSetExample for custom mapped namespace location:

```
[root@xxxxx lib]# hdfs dfs -ls /tmp/rohit
Found 3 items
drwxr-xr-x   - cdap supergroup          0 2016-07-17 23:26 /tmp/rohit/artifacts
drwxr-xr-x   - cdap supergroup          0 2016-07-17 23:26 /tmp/rohit/data
drwxr-xr-x   - cdap supergroup          0 2016-07-17 23:26 /tmp/rohit/programs
[root@xxxx lib]# hdfs dfs -ls /tmp/rohit/data
Found 2 items
drwxr-xr-x   - cdap supergroup          0 2016-07-17 23:28 /tmp/rohit/data/counts
drwxr-xr-x   - cdap supergroup          0 2016-07-17 23:26 /tmp/rohit/data/example
[root@xxx lib]# hdfs dfs -ls /tmp/rohit/data
Found 2 items
drwxr-xr-x   - cdap supergroup          0 2016-07-17 23:28 /tmp/rohit/data/counts
drwxr-xr-x   - cdap supergroup          0 2016-07-17 23:26 /tmp/rohit/data/example
[root@xxxx lib]# hdfs dfs -ls /tmp/rohit/data/counts
Found 1 items
drwxr-xr-x   - cdap supergroup          0 2016-07-17 23:28 /tmp/rohit/data/counts/counts.out
[root@xxx lib]#
```

Commit:
Fix unit tests
